pkg/test/framework/api/component/descriptor.go

```diff
@@ -14,37 +14,47 @@
 
 package component
 
-import "fmt"
+import (
+	"fmt"
+)
 
 var _ Requirement = &Descriptor{}
 
-// Variant allows an environment to support multiple components with the same ID. For example, an environment might
-// support two Widget components: "fake widget" and "real widget".
+// Variant supports multiple components with the same ID. For example, a test may use two Widget
+// components: "fake widget" and "real widget".
 type Variant string
 
+// CreateDescriptor creates a descriptor with the given ID and variant.
+func NewDescriptor(id ID, variant string) *Descriptor {
+	return &Descriptor{Key: Key{ID: id, Variant: Variant(variant)}}
+}
+
 // Descriptor describes a component of the testing framework.
 type Descriptor struct {
-	ID
+	Key               Key
 	IsSystemComponent bool
-	Variant           Variant
+	Configuration     Configuration
```

Comment:
Just for my own understanding ... Descriptor now seems to be a pallet of options. The fields may or may not be filled out when a component is being required, so the system will have to sort out the missing pieces (e.g. `IsSystemComponent`). I assume it's doing this via the default for the component ID? If there is an existing component available, do we re-configure in that case?

Reply:
Yeah this was a bit hard to decide on, mostly around requires. The model in this PR is that a variant Descriptor will override the default Descriptor (with an "" variant). For Configuration and Requires the default values will be used if they are not overridden (this means you can't remove all requirements except by replacing with something like ids.Environment). I didn't do that for IsSystemComponent, it seems like that should be explicitly set. But I haven't worked through the implications with Kube environment for that yet.

Actually I just realized I don't really always pull in the default for the component ID, that's a bit ad-hoc. Let me add a test for that and verify it works correctly, will ping PR when that is in.

There is also weirdness where if the user requests a component by ID, that means any instance of that ID, not a specific variant. That is necessary when resolving requirements, so for example if you require ids.Pilot and there is a variant of pilot running, we don't want to spin up both the variant and the default version.

I think it actually works OK, if you want a specific variant you can use a Descriptor (which we can pre-create for variants like we do with the environment variants). So in practice this looks like the right model.